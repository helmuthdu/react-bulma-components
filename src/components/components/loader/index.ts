import './loader.scss';

export * from './loader';
