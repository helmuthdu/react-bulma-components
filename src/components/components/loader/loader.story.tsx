import { storiesOf } from '@storybook/react';
import * as React from 'react';
import { Loader } from '.';

storiesOf('Loader', module)
  .add('Default', () => <Loader />)
  .add('with overlay', () => <Loader overlay />);
