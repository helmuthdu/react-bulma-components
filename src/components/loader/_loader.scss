$loading-background-legacy: #7f7f7f !default;
$loading-background: rgba(255, 255, 255, 0.5) !default;
$loading-background-inverted: rgba(0, 0, 0, 0.6) !default;
$loading-icon-size: 3em !default;
$loading-full-page-icon-size: 5em !default;

.loader {
  @include overlay;
  align-items: center;
  display: flex;
  justify-content: center;
  overflow: hidden;
  opacity: 0;
  visibility: hidden;
  transition: opacity 600ms, visibility 600ms;

  &.is-active {
    opacity: 1;
    visibility: visible;
  }

  &.is-fullscreen {
    z-index: 999;
    position: fixed;

    .loading-icon {
      &:after {
        top: calc(50% - #{$loading-full-page-icon-size / 2});
        left: calc(50% - #{$loading-full-page-icon-size / 2});
        width: $loading-full-page-icon-size;
        height: $loading-full-page-icon-size;
      }
    }
  }

  .loader-background {
    @include overlay;
    animation-duration: 0.5s;
    transition: background-color 0.5s linear;
    background: $loading-background-legacy;
    background: $loading-background;

    &.is-inverted {
      background-color: $loading-background-inverted;
    }
  }

  .loader-icon {
    position: relative;

    &:after {
      @include loader;
      position: absolute;
      top: calc(50% - #{$loading-icon-size / 2});
      left: calc(50% - #{$loading-icon-size / 2});
      width: $loading-icon-size;
      height: $loading-icon-size;
      border-width: 0.25em;
    }
  }
}
