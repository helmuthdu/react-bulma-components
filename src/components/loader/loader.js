import cn from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import modifiers from '../../modifiers';
import { Element } from '../element';

export const Loader = ({ overlay, onClick, className, ...props }) => (
  <Element
    {...props}
    className={cn('loader', className, {
      'is-overlay': overlay
    })}
  >
    <div className="loader-background" onClick={onClick} />
    <div className="loader-icon" />
  </Element>
);

Loader.propTypes = {
  ...modifiers.propTypes,
  onClick: PropTypes.func,
  className: PropTypes.string,
  style: PropTypes.object,
  renderAs: PropTypes.oneOfType([PropTypes.string, PropTypes.func])
};

Loader.defaultProps = {
  ...modifiers.defaultProps,
  overlay: false,
  renderAs: 'div'
};
