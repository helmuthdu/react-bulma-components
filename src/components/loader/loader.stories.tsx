import { boolean } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import * as React from 'react';
import { Loader } from '.';

storiesOf('Loader', module).add('Default', () => (
  <div style={{ position: 'relative', width: '100%', height: '200px', backgroundColor: '#f1f1f1' }}>
    <Loader
      active={boolean('Active', false)}
      inverted={boolean('Inverted', false)}
      fullscreen={boolean('Fullscreen', false)}
    />
  </div>
));
