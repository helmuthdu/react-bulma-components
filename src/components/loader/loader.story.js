import { storiesOf } from '@storybook/react';
import React from 'react';
import { Loader } from '.';

storiesOf('Loader', module)
  .add('Default', () => <Loader />)
  .add('with overlay', () => <Loader overlay />);
