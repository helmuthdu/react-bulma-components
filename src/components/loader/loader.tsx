import cn from 'classnames';
import * React from 'react';
import modifiers, { ElementModifier } from '../../modifiers';
import { Element } from '../../elements';

type LoaderProps = ElementModifier & {
  onClick?: (...args: any[]) => any;
};

export const Loader: React.FunctionComponent<LoaderProps> = ({
  overlay,
  onClick,
  className,
  ...props
}: LoaderProps) => (
  <Element
    {...props}
    className={cn('loader', className, {
      'is-overlay': overlay
    })}
  >
    <div className="loader-background" onClick={onClick} />
    <div className="loader-icon" />
  </Element>
);

Loader.defaultProps = {
  ...modifiers.defaultProps,
  overlay: false
};
