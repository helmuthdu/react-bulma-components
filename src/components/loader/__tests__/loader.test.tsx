import { render } from '@testing-library/react';
import * as React from 'react';
import { Loader } from '../loader';

describe('Loader component', () => {
  it('should render', () => {
    const { asFragment } = render(<Loader />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should concat classname in props with classname', () => {
    const { container } = render(<Loader className="other-class test" />);
    expect(container.querySelector('.loader')).toHaveClass('other-class', 'test');
  });

  it('should be active', () => {
    const { container } = render(<Loader active />);
    expect(container.querySelector('.loader')).toHaveClass('is-active');
  });

  it('should have inverted background color', () => {
    const { container } = render(<Loader inverted />);
    expect(container.querySelector('.loader-background')).toHaveClass('is-inverted');
  });

  it('should render in fullscreen', () => {
    const { container } = render(<Loader active fullscreen />);
    expect(container.querySelector('.loader')).toHaveClass('is-active', 'is-fullscreen');
  });

  it('should have custom inline styles', () => {
    const { asFragment } = render(<Loader as="section" style={{ width: 200, zIndex: 1 }} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
