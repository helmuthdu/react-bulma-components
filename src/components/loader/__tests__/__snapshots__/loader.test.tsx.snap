// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader component should have custom inline styles 1`] = `
<DocumentFragment>
  <section
    class="loader"
    style="width: 200px; z-index: 1;"
  >
    <div
      class="loader-background"
    />
    <div
      class="loader-icon"
    />
  </section>
</DocumentFragment>
`;

exports[`Loader component should render 1`] = `
<DocumentFragment>
  <div
    class="loader"
  >
    <div
      class="loader-background"
    />
    <div
      class="loader-icon"
    />
  </div>
</DocumentFragment>
`;
